doctest/extensions/doctest_mpi.h

```diff
@@ -2,80 +2,104 @@
 
 #ifdef DOCTEST_CONFIG_IMPLEMENT
 
-#include "doctest/doctest.h"
+#include "doctest/extensions/mpi_sub_comm.h"
 #include "mpi_reporter.h"
-
-#else
-
-#include "mpi.h"
-#include <numeric>
-#include <vector>
-#include "doctest/doctest.h"
-#include <cassert>
-#include <string>
+#include <unordered_map>
 
 namespace doctest {
 
-inline
-int mpi_world_nb_procs() {
-  int n;
-  MPI_Comm_size(MPI_COMM_WORLD, &n);
-  return n;
+// Each time a MPI_TEST_CASE is executed on N procs,
+// we need a sub-communicator of N procs to execute it.
+// It is then registered here and can be re-used
+// by other tests that requires a sub-comm of the same size
+std::unordered_map<int,mpi_sub_comm> sub_comms_by_size;
+
+// Record if at least one MPI_TEST_CASE was registered "skipped"
+// because there is not enought procs to execute it
+int nb_test_cases_skipped_insufficient_procs = 0;
+
+// Can be safely called before MPI_Init()
+//   This is needed for MPI_TEST_CASE because we use doctest::skip()
+//   to prevent execution of tests where there is not enough procs,
+//   but doctest::skip() is called during test registration, that is, before main(), and hence before MPI_Init()
+int
+mpi_comm_world_size() {
+  #if defined(OPEN_MPI)
+    const char* size_str = std::getenv("OMPI_COMM_WORLD_SIZE");
+  #elif defined(I_MPI_VERSION) || defined(MPI_VERSION)
+    const char* size_str = std::getenv("PMI_SIZE"); // see https://community.intel.com/t5/Intel-oneAPI-HPC-Toolkit/Environment-variables-defined-by-intel-mpirun/td-p/1096703
+  #else
+    #error "Unknown MPI implementation"
```

Comment:
I guess OpenMPI and IntelMPI are the two most commonly used, but we should keep our eyes open also to support others like MPICH

Reply:
another idea:
We could optionally ask the user for a custom env-var in case we don't directly support the MPI-implementation
i.e. sth on the line of the following:
~~~c++
for (const auto& env_var_name : {"DOCTEST_COMM_WORLD_SIZE", "OMPI_COMM_WORLD_SIZE", "PMI_SIZE"})  { // can be easily extended if we know the env-var names from other MPI-implementations
    const char* size_str = std::getenv(env_var_name);
    if (size_str) {
        int size = std::stoi(size_str);
        assert (size > 0);
        return size;
    }
}

error: please define "DOCTEST_COMM_WORLD_SIZE" to match the number of cores running with
~~~

and then inside `doctest::mpi_init_thread()` check if it is consistent with what is returned from `MPI_Comm_size`

Reply:
Yes I would like to support other implementations but I am not using them personally (and installing MPI is not something I particularly long for). I will definitely upstream a patch if I ever get my hands on one of them.
Regarding your `DOCTEST_COMM_WORLD_SIZE` suggestion, I was going to implement it, but I think that we are better off failing with a clear message that encourages to upstream the problem rather than silently allowing people to get around with an incomplete support. Plus, I am wondering if `PMI_SIZE `isn't the name for MPICH and MVAPICH anyway since they all share a common root. What do you think ?